proposals/20251127-thread-safe-sinsp-thread-manager.md

```diff
@@ -0,0 +1,257 @@
+   * **Writers:** use RCU pointer assignment *rcu_assign_pointer()* for atomic pointer updates, and a mutex to serialize writers when applying topological changes. Note that it may be possible in theory to have a lock-free writer, based on CAS operations, but the complexity of the solution would increase tremendously, and it would require a weaker consistency model.
+   * **Reclamation:** retire object (Asynchronous deletion).
+
+### Data Structures
+
+#### The Thread Object (sinsp\_thread\_info)
+
+The object carries its own navigation pointers to avoid allocating separate list nodes for iteration.
+The following is a simplified version of the actual structure.
+
+```c++
+struct sinsp_thread_info {
+
+    // --- Domain Data ---
+    // Immutable after creation, or updated via atomic operations or other
+    // synchronization mechanisms.
+    int64_t m_tid;
+    int64_t m_tgid;
+    std::string m_comm;
+    // ... other fields ...
+
+    // Topology Pointers used for process tree transversal, re-parenting, etc.
+    std::atomic<sinsp_thread_info*> m_ptid{nullptr}; // parent thread pointer
+    list_node<sinsp_thread_info> m_children; // intrusive list of children
+
+    // Intrusive list node for fast thread group leader traversal
+    list_head m_tasks{nullptr};
+
+    sinsp_thread_info(int64_t t) : tid(t) {}
+};
+```
+
+In the pseudo-code above, `list_head` is an intrusive RCU-protected linked list node, similar to the Linux kernel's [list_head](https://github.com/torvalds/linux/blob/559e608c46553c107dbba19dae0854af7b219400/include/linux/types.h#L200-L202) struct.
+It provides the necessary functionality to link `sinsp_thread_info` objects into various lists while ensuring thread-safe traversal and modification using RCU primitives.
+
+### The Manager (sinsp\_thread\_manager)
+
+The simplified version of the manager structure is as follows:
+
+```c++
+class sinsp_thread_manager {
+private:
+    // Store threads for O(1) lookup by TID
+    std::vector<std::atomic<sinsp_thread_info*>> table_;
+
+    // Serialize writers
+    std::mutex process_list_lock;
+    std::atomic<sinsp_thread_info*> list_head_{nullptr};
+
+public:
+    // PID_MAX is typically 32768, but can be up to 4M on 64-bit systems.
+    explicit sinsp_thread_manager(size_t max_pids = (1 << 22))
+        : table_(max_pids) {};
+};
+```
+
+### Concurrency Model
+
+#### The Hot Path: Lookup (Wait-Free/Lock-Free)
+
+This should be the most frequent operation, executed during event parsing and rules evaluation. The RCU primitives ensure that multiple readers can access the data in a lock-free and wait-free manner.
+
+Using a visitor function to access the *sinsp\_thread\_info* would be a good approach to be able to keep control over the reader critical section. Alternatively, we could return a smart pointer based on RCU primitives.
+
+```c++
+using thread_visitor = std::function<void(const sinsp_thread_info*)>;
```

Comment:
Here we used `std::function`, but since this is the hot path, I would like to evaluate function pointers as a more performant alternative. If we find out that the performance benefit is worthy in this latter case but, at the same time, we really need to use functors, we could evaluate providing two function signatures: one with pointers and one with `std::function`.